### PR TITLE
feat: simplify off-chain-auth by removing nonce and using ed25519 to verify signature

### PR DESCRIPTION
**Pull request by ruojunm**
### Description

feat: simplify off-chain-auth by removing nonce and using ed25519 to verify signature

### Rationale

off-chain-auth was over-designed and now we need make it simpler and easy to use. 

### Example

- Register
```
curl --location --request POST '127.0.0.1:9033/auth/update_key_v2' \
--header 'Referer: https://dcellar-qa.fe.nodereal.cc/' \
--header 'x-gnfd-user-address: 0xA4cFe2dE3e45C043524aaC46fDdFb46311aF0af6' \
--header 'x-gnfd-app-domain: https://dcellar-qa.fe.nodereal.cc' \
--header 'x-gnfd-app-reg-public-key: 5a3271a38dfadf7b73769173d03ac6d2442f7015bf2224cfa15f8a65008cdf8a' \
--header 'x-gnfd-expiry-timestamp: 2024-03-04T09:07:48Z' \
--header 'authorization: GNFD1-ETH-PERSONAL_SIGN,SignedMsg=https://dcellar-qa.fe.nodereal.cc wants you to sign in with your BNB Greenfield account:0xA4cFe2dE3e45C043524aaC46fDdFb46311aF0af6\nRegister your identity public key 5a3271a38dfadf7b73769173d03ac6d2442f7015bf2224cfa15f8a65008cdf8a\nURI: https://dcellar-qa.fe.nodereal.cc\nVersion: 1\nChain ID: 9000\nIssued At: 2024-02-28T09:07:48Z\nExpiration Time: 2024-03-04T09:07:48Z,Signature=0x827ad691f2e0a97c5bccf8592f5c62d94950797b5a8505089324190a8358eb6c2d7adbd254c89f27a2135a5680658badda3d4bcfb93118f24a801b59259d380c1c' \
--header 'Origin: https://dcellar-qa.fe.nodereal.cc' 

```
- Adding signature by using ed25519
```
curl --location 'https://clyde-test8-2.gf-devnet-sp-a-bk.dev.nodereal.cc/clyde-test8-2.txt' \
--header 'User-Agent: Greenfield (darwin; arm64) greenfield-go-sdk/v0.1.0' \
--header 'X-Gnfd-Date: 2024-03-05T07:45:58Z' \
--header 'x-gnfd-app-domain: https://test.domain.com' \
--header 'authorization: GNFD2-EDDSA,Signature=784831a5d1f588090295c1f64ef43b003effff93d092f04f88c67946af108fd7f7abe6180865e270a1a34a8f46d1c5f8c13b7923708c285523dc222b61c3140b' \
--header 'Content-Type: application/octet-stream' \
--header 'X-Gnfd-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855' \
--header 'x-gnfd-expiry-timestamp: 2024-03-05T08:02:38Z' \
--header 'X-Gnfd-User-Address: 0xA4cFe2dE3e45C043524aaC46fDdFb46311aF0af6' \
--header 'X-Gnfd-App-Reg-Public-Key: e0d61609201e9cece55e8999a40a74f5119003b1d49fad32882fa50547c4bf90'
```

### Changes

Notable changes: 
* removing nonce concept during public key registration process
* use ed25519 rather than bn254 (in gnark-crypto) for eddsa curve
* support multiple keys for each "user + domain" combination
* new authentication method is called GNFD2-EDDSA

### Potential Impacts
* No impacts for clients that using old version of off-chain-auth, as the changes are backward compatible.

### CODE REVIEW

File: base/gfspclient/helper_test.go [hunk 1]
```diff
@@ -145,6 +145,31 @@ func (*mockApproverServer) GfSpAskApproval(ctx context.Context, req *gfspserver.
 
 type mockAuthenticatorServer struct{}
 
+func (s mockAuthenticatorServer) GetAuthKeyV2(ctx context.Context, request *gfspserver.GetAuthKeyV2Request) (*gfspserver.GetAuthKeyV2Response, error) {
+	//TODO implement me
+	panic("implement me")
```
Comment (mliu233) [inline base/gfspclient/helper_test.go:150]: Fill in this part after gfspserver server part got implemented?

File: modular/gater/auth_handler.go [hunk 2]
```diff
@@ -342,3 +598,29 @@ func (g *GateModular) verifySignedContent(signedContent string, expectedDomain s
 
 	return nil
 }
+
+func (g *GateModular) verifySignedContentV2(signedContent string, expectedDomain string, expectedPublicKey string, expectedExpiryDate string) error {
+	pattern := `(.+) wants you to sign in with your BNB Greenfield account:\n*(.+)\n*Register your identity public key (.+)\n*URI: (.+)\n*Version: (.+)\n*Chain ID: (.+)\n*Issued At: (.+)\n*Expiration Time: (.+)`
```
Comment (mliu233) [inline modular/gater/auth_handler.go:603]: Consider make this pattern a constant if we do not change it often?

File: modular/gater/migrate_handler_test.go [hunk 3]
```diff
@@ -10,14 +10,15 @@ import (
 	"testing"
 	"time"
 
+	"github.com/gorilla/mux"
+	"github.com/stretchr/testify/assert"
+	"go.uber.org/mock/gomock"
+
```
Comment (mliu233) [inline modular/gater/migrate_handler_test.go:16]: restore imports here

File: .github/workflows/docker-develop.yml [hunk 4]
```diff
@@ -2,7 +2,7 @@ name: Docker-CI
 
 on:
   push:
-    branches: [ develop, master ]
+    branches: [ develop, master, off-chain-auth-v2]
```
Comment (constwz) [inline .github/workflows/docker-develop.yml:5]: need delete off-chain-auth-v2

File: base/gfspclient/helper_test.go (hunk 1 above)
Comment (ruojunm) [reply base/gfspclient/helper_test.go]: fixed

File: modular/gater/auth_handler.go (hunk 2 above)
Comment (ruojunm) [reply modular/gater/auth_handler.go:603]: Done

File: modular/gater/migrate_handler_test.go (hunk 3 above)
Comment (ruojunm) [reply modular/gater/migrate_handler_test.go:16]: Done

File: .github/workflows/docker-develop.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (ruojunm) [reply .github/workflows/docker-develop.yml:5]: Done

File: modular/gater/auth_handler.go [hunk 5]
```diff
@@ -227,6 +236,253 @@ func (g *GateModular) updateUserPublicKeyHandler(w http.ResponseWriter, r *http.
...
+	}
+
+	// validate domain header
+	if domain == "" {
+		log.CtxErrorw(reqCtx.Context(), "failed to check GnfdOffChainAuthAppDomainHeader header")
+		err = ErrInvalidDomainHeader
+		return
+	}
+
+	ctx := log.Context(context.Background(), account, domain)
```
[the remark this answers is not in the document]
Comment (ruojunm) [reply modular/gater/auth_handler.go:388]: Done

File: modular/authenticator/authenticator.go [hunk 6]
```diff
@@ -164,6 +171,102 @@ func (a *AuthenticationModular) VerifyGNFD1EddsaSignature(ctx context.Context, a
...
+		UserAddress:  account,
+		Domain:       domain,
+		PublicKey:    publicKey,
+		ExpiryDate:   time.UnixMilli(expiryDate),
+		CreatedTime:  time.Now(),
+		ModifiedTime: time.Now(),
+	}
+	err := a.baseApp.GfSpDB().InsertAuthKeyV2(newRecord)
+	if err != nil {
+		log.CtxErrorw(ctx, "failed to UpdateUserPublicKeyV2 when saving key")
```
[the remark this answers is not in the document]
Comment (ruojunm) [reply modular/authenticator/authenticator.go:235]: Sure. Done

File: modular/gater/auth_handler.go [hunk 7]
```diff
@@ -227,6 +236,253 @@ func (g *GateModular) updateUserPublicKeyHandler(w http.ResponseWriter, r *http.
...
+
+	// ignore the error, because the listUserPublicKeyV2 does not need signature
+	reqCtx, _ = NewRequestContext(r, g)
+
+	account := reqCtx.request.Header.Get(GnfdUserAddressHeader)
+	domain := reqCtx.request.Header.Get(GnfdOffChainAuthAppDomainHeader)
+
+	// validate account header
+	if ok := common.IsHexAddress(account); !ok {
+		log.Errorw("failed to check account address", "account_address", account, "error", err)
```
[the remark this answers is not in the document]
Comment (ruojunm) [reply modular/gater/auth_handler.go:376]: Fixed

File: modular/gater/request_context.go [hunk 8]
```diff
@@ -282,6 +292,50 @@ func (r *RequestContext) verifyGNFD1EddsaSignatureFromPreSignedURL(authenticatio
...
+	}
+
+	// check request integrity
+	realMsgToSign := commonhttp.GetMsgToSignInGNFD1AuthForPreSignedURL(r.request)
+	_, err = r.g.baseApp.GfSpClient().VerifyGNFD2EddsaSignature(r.Context(), account, domain, userPublicKey, offChainSig, realMsgToSign)
+	if err != nil {
+		log.Errorw("failed to verify off chain signature", "error", err)
+		return nil, err
+	} else {
+		userAddress, _ := sdk.AccAddressFromHexUnsafe(account)
```
[the remark this answers is not in the document]
Comment (ruojunm) [reply modular/gater/request_context.go:310]: fixed

**Review by constwz — APPROVED** [review]
LGTM

**Review by alexgao001 — APPROVED** [review]

**Review by zjubfd — APPROVED** [review]

**Pull request merged**